src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -217,6 +226,10 @@ public class ModelInterpreter implements Serializable {
     def dockerOrWithout(Agent agent, Closure body) {
         if (agent.docker != null) {
             return {
+                script.stage(SyntheticStage.dockerPull()) {
+                    Utils.markSyntheticStage(SyntheticStage.dockerPull(), SyntheticStage.SYNTHETIC_PRE)
+                    script.getProperty("docker").image(agent.docker).pull()
+                }
                 script.getProperty("docker").image(agent.docker).inside {
```

Comment:
So at least at some point in the past, that very much didn't work. Dunno if it would now, I'll give it a try.